File: app/server/notify/channels/bark.ts
/**
 * Bark 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult, NotifyParams } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'bark' });

export const barkChannel: NotifyChannel = {
    name: 'bark',
    enabled: false,

    async send(text: string, desp: string, params: NotifyParams = {}): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('BARK_PUSH')) {
                resolve({ success: false, message: 'BARK_PUSH 未配置' });
                return;
            }

            let BARK_PUSH = getConfig('BARK_PUSH') || '';
            const BARK_ICON = getConfig('BARK_ICON');
            const BARK_SOUND = getConfig('BARK_SOUND');
            const BARK_GROUP = getConfig('BARK_GROUP');
            const BARK_LEVEL = getConfig('BARK_LEVEL');
            const BARK_ARCHIVE = getConfig('BARK_ARCHIVE');
            const BARK_URL = getConfig('BARK_URL');

            // 兼容BARK本地用户只填写设备码的情况
            if (!BARK_PUSH.startsWith('http')) {
                BARK_PUSH = `https://api.day.app/${BARK_PUSH}`;
            }

            const body = {
                title: text,
                body: desp,
                icon: BARK_ICON,
                sound: BARK_SOUND,
                group: BARK_GROUP,
                isArchive: BARK_ARCHIVE,
                level: BARK_LEVEL,
                url: BARK_URL,
                ...params,
            };

            $.post({ url: BARK_PUSH, json: body }, (err, _resp, data) => {
                if (err) {
                    logger.error({ err }, 'Bark 发送失败');
                    resolve({ success: false, error: err as Error });
                } else {
                    const result = data as { code?: number; message?: string };
                    if (result?.code === 200) {
                        logger.info('Bark 发送成功');
                        resolve({ success: true, message: '发送成功' });
                    } else {
                        logger.warn({ body: data }, 'Bark 发送失败');
                        resolve({ success: false, message: result?.message || '发送失败' });
                    }
                }
            });
        });
    }
};

// 检查是否启用
if (hasConfig('BARK_PUSH')) {
    barkChannel.enabled = true;
}

export default barkChannel;

File: app/server/internal/notify/channels/bark.go
package channels

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type Bark struct {
	enabled bool
}

func (c *Bark) Name() string            { return "bark" }
func (c *Bark) Enabled() bool           { return c.enabled }
func (c *Bark) SetEnabled(enabled bool) { c.enabled = enabled }

func (c *Bark) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("BARK_PUSH") {
		return notify.NotifyResult{Success: false, Message: "BARK_PUSH 未配置"}
	}

	barkPush := notify.GetConfig("BARK_PUSH")
	if !strings.HasPrefix(barkPush, "http") {
		barkPush = "https://api.day.app/" + barkPush
	}

	body := map[string]string{
		"title":     text,
		"body":      desp,
		"icon":      notify.GetConfig("BARK_ICON"),
		"sound":     notify.GetConfig("BARK_SOUND"),
		"group":     notify.GetConfig("BARK_GROUP"),
		"isArchive": notify.GetConfig("BARK_ARCHIVE"),
		"level":     notify.GetConfig("BARK_LEVEL"),
		"url":       notify.GetConfig("BARK_URL"),
	}

	resp, err := notify.HTTPPost(barkPush, notify.HttpRequestOptions{JSON: body})
	if err != nil {
		slog.Error("Bark 发送失败", "err", err)
		return notify.NotifyResult{Success: false, Message: fmt.Sprintf("发送失败: %v", err), Err: err}
	}

	var result struct {
		Code    float64 `json:"code"`
		Message string  `json:"message"`
	}
	if err := json.Unmarshal([]byte(resp.Body), &result); err != nil {
		slog.Error("Bark 解析响应失败", "err", err)
		return notify.NotifyResult{Success: false, Message: "解析响应失败", Err: err}
	}

	if result.Code == 200 {
		slog.Info("Bark 发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	msg := result.Message
	if msg == "" {
		msg = "发送失败"
	}
	slog.Warn("Bark 发送失败", "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: msg}
}

func init() {
	ch := &Bark{}
	if notify.HasConfig("BARK_PUSH") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
